### PR TITLE
Replace usage of deprecated gtk::Dialog with adw::MessageDialog in the books listings.

### PR DESCRIPTION
**Pull request by zekefast**
# Overview

[`gtk::Dialog` was deprecated as of GTK 4.10](https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gtk/deprecated/gtkdialog.c#L173) and documentation proposes to use `gtk::Window` instead, but I found [`libadwaita::MessageDialog`](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.3/class.MessageDialog.html) which were added in `1.2` version of `libadwaita` and could be a nice replacement for usage of `gtk::Dialog` in the book.

# Changes
- Bump version requirement in `Cargo.toml` file for `libadwaita` dependency from `*` to `>= 0.3.1` in order to be able to enable `v1_2` feature.
  `0.3.1` is the latest version of libadwaita as ATM.
  `v1_2` is required to be able to use `adw::MessageDialog` wich were added in `1.2` of `libadwaita`.
- Adjust imports:
  - Remove not used any more `gtk::{Dialog, DialogFlags, ResponseType}`
  - Add `adw::{MessageDialog, ResponseAppearance}`
- As `adw::MessageDialog` does not provide any predefined responses it was introduced new ones in `Window::new_collection` methods using the following local variables:
  - `cancel_response`
  - `create_response`
- Creation of `entry` was moved at the beginning of `new_collection` method in order to be able to use it later in `adw::MessageDialogBuilder.extra_child` method.
- Buttons creation were replaced with usage of `adw::MessageDialog.add_responses` method.
- Set of insetnsitivity of "Create" button were done through usage of `adw::MessageDialog.set_response_enabled` method.
- Set suggested action through `adw::MessageDialog.set_response_appearance` method.
- Make small adjustments regarding APIs difference of between `gtk::Dialog` and `adw::MessageDialog` for the following methods:
  - `connect_response`

# TODO
- [x] Record animations with new "New Collection" window and add them to this PR.

### CODE REVIEW

**Comment by zekefast** [comment]
@Hofer-Julian This is how new dialog looks like

Screenshot 2023-03-25 16-06-45

**Comment by zekefast** [comment]
@Hofer-Julian if changes are fine with you and you gonna merge the proposal could you please give me background image to record animations (in order to make them on the same background as the rest of them) and may be saved json for task and collections (if you have one).

Can you suggest some tools used for image recording and cropping?

File: book/listings/todo/8/window/mod.rs [hunk 1]
```diff
@@ -397,46 +397,38 @@ impl Window {
...
-        );
-        dialog.set_default_response(ResponseType::Accept);
-
-        // Make the dialog button insensitive initially
-        let dialog_button = dialog
-            .widget_for_response(ResponseType::Accept)
-            .expect("The dialog needs to have a widget for response type `Accept`.");
-        dialog_button.set_sensitive(false);
+        const CANCEL_RESPONSE: &str = "cancel";
+        const CREATE_RESPONSE: &str = "create";
```
Comment (Hofer-Julian) [inline book/listings/todo/8/window/mod.rs:401]:
```suggestion
        let cancel_response = "cancel";
        let create_response = "create";
```

File: book/listings/todo/8/window/mod.rs [hunk 2]
```diff
@@ -397,46 +397,38 @@ impl Window {
...
-            .margin_top(12)
-            .margin_bottom(12)
-            .margin_start(12)
-            .margin_end(12)
             .placeholder_text("Name")
             .activates_default(true)
             .build();
-        dialog.content_area().append(&entry);
+
+        // Create new Dialog
```
Comment (Hofer-Julian) [inline book/listings/todo/8/window/mod.rs:409]:
```suggestion
        // Create new dialog
```

File: book/listings/todo/8/window/mod.rs [hunk 3]
```diff
@@ -397,46 +397,38 @@ impl Window {
...
+        let dialog = MessageDialog::builder()
+            .heading("New Collection")
+            .transient_for(self)
+            .modal(true)
+            .destroy_with_parent(true)
+            .close_response(CANCEL_RESPONSE)
+            .default_response(CREATE_RESPONSE)
+            // add entry to the dialog
```
Comment (Hofer-Julian) [inline book/listings/todo/8/window/mod.rs:417]:
```suggestion
```

File: book/listings/todo/8/window/mod.rs [hunk 4]
```diff
@@ -397,46 +397,38 @@ impl Window {
...
         // Set entry's css class to "error", when there is no text in it
         entry.connect_changed(clone!(@weak dialog => move |entry| {
             let text = entry.text();
-            let dialog_button = dialog.
-                widget_for_response(ResponseType::Accept).
-                expect("The dialog needs to have a widget for response type `Accept`.");
             let empty = text.is_empty();
 
-            dialog_button.set_sensitive(!empty);
+            dialog.set_response_enabled("create", !empty);
```
Comment (Hofer-Julian) [inline book/listings/todo/8/window/mod.rs:431]:
```suggestion
            dialog.set_response_enabled(create_response, !empty);
```

File: book/listings/todo/8/window/mod.rs [hunk 5]
```diff
@@ -447,12 +439,13 @@ impl Window {
 
         // Connect response to dialog
         dialog.connect_response(
+            None,
             clone!(@weak self as window, @weak entry => move |dialog, response| {
                 // Destroy dialog
                 dialog.destroy();
 
                 // Return if the user chose a response different than `Accept`
```
Comment (Hofer-Julian) [inline book/listings/todo/8/window/mod.rs:447]:
```suggestion
                // Return if the user chose a response different than `create_response`
```

**Review by Hofer-Julian — COMMENTED** [review]
Thanks, @zekefast, that looks indeed like a good use case for `adw::MessageDialog`.

It took me a bit to find it, but you can find the wallpaper here: https://gitlab.gnome.org/GNOME/gnome-backgrounds/-/blob/42.0/backgrounds/truchet-l.jpg

For recording, GNOME has that built in.
You only have to press the PRINT key.
Otherwise, https://github.com/SeaDve/Kooha works too.

**Comment by Hofer-Julian** [comment]
You can use this as a starting point:

```json
[
    {
        "title": "Collection Number One",
        "tasks_data": [
            {
                "completed": false,
                "content": "Task Number One"
            },
            {
                "completed": false,
                "content": "Task Number Two"
            },
            {
                "completed": false,
                "content": "Task Number Three"
            },
            {
                "completed": false,
                "content": "Task Number Four"
            },
            {
                "completed": false,
                "content": "Task Number Five"
            },
            {
                "completed": false,
                "content": "Task Number Six"
            },
            {
                "completed": false,
                "content": "Task Number Seven"
            },
            {
                "completed": false,
                "content": "Task Number Eight"
            }
        ]
    }
]
```

**Comment by zekefast** [comment]
@Hofer-Julian I updated first video. I hope you don't mind using VP9 instead of VP8. I'll prepare second one tomorrow. Meanwhile you could check if its Okay.

**Comment by Hofer-Julian** [comment]
> @Hofer-Julian I updated first video. I hope you don't mind using VP9 instead of VP8. I'll prepare second one tomorrow. Meanwhile you could check if its Okay.

The video looks fine.

**Comment by sdroege** [comment]
Isn't libadwaita meant for GNOME applications specifically? It would seem strange to use it here then.

**Comment by Hofer-Julian** [comment]
> Isn't libadwaita meant for GNOME applications specifically? It would seem strange to use it here then.

The chapter that uses a dialog is about libadwaita so it is fine

**Comment by zekefast** [comment]
@Hofer-Julian I updated `src/vid/todo_8_adaptive_sidebar.webm`. Now all videos with dialog window should show new `adw::MessageDialog`.

**Comment by Hofer-Julian** [comment]
Thanks!
Now only the text in the book needs to be adapted :)

**Comment by zekefast** [comment]
> Now only the text in the book needs to be adapted :)

Which one? I tried to look through, but I haven't found direct mentions of `gtk::Dialog` usage. There are references like "dialog" but it seems fine.

Or do you want to extend explanation of the chapter regarding usage of `adw::MessageDialog`?

**Comment by Hofer-Julian** [comment]
> > Now only the text in the book needs to be adapted :)
> 
> Which one? I tried to look through, but I haven't found direct mentions of `gtk::Dialog` usage. There are references like "dialog" but it seems fine.
> 
> Or do you want to extend explanation of the chapter regarding usage of `adw::MessageDialog`?

Looking at it more closely, you are right, the text is still valid.
Maybe change 

```diff
- Here, we create the dialog, set up the buttons as well as add the entry to it.
+ Here, we create the [`adw::MessageDialog`](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/struct.MessageDialog.html), set up the buttons as well as add the entry to it. 
```

That should make it easy to see that `MessageDialog comes from libadwaita.
Also this should be adapted:

```diff
- cargo add libadwaita --rename adw
+ cargo add libadwaita --rename adw --features v1_2
```

File: book/listings/todo/8/window/mod.rs [hunk 6]
```diff
@@ -397,46 +397,37 @@ impl Window {
...
-
-        // Make the dialog button insensitive initially
-        let dialog_button = dialog
-            .widget_for_response(ResponseType::Accept)
-            .expect("The dialog needs to have a widget for response type `Accept`.");
-        dialog_button.set_sensitive(false);
+        let cancel_response = "cancel";
+        let create_response = "create";
 
-        // Create entry and add it to the dialog
```
Comment (Hofer-Julian) [inline book/listings/todo/8/window/mod.rs:420]: Why did you change this comment?

File: book/listings/todo/8/window/mod.rs [hunk 7]
```diff
@@ -397,46 +397,37 @@ impl Window {
...
-        );
-        dialog.set_default_response(ResponseType::Accept);
-
-        // Make the dialog button insensitive initially
-        let dialog_button = dialog
-            .widget_for_response(ResponseType::Accept)
-            .expect("The dialog needs to have a widget for response type `Accept`.");
-        dialog_button.set_sensitive(false);
+        let cancel_response = "cancel";
+        let create_response = "create";
```
Comment (Hofer-Julian) [inline book/listings/todo/8/window/mod.rs:401]: I feel this should be declared after `Entry`

File: book/listings/todo/8/window/mod.rs (hunk 6 above)
Comment (zekefast) [reply book/listings/todo/8/window/mod.rs:420]: I split it to two parts one for creation, other for addition to dialog as addition now is made using `extra_child` in dialog builder and detached from creation code.

And later you suggested to get rid of addition comment [here](https://github.com/gtk-rs/gtk4-rs/pull/1343#discussion_r1148420929).

Comment (Hofer-Julian) [reply book/listings/todo/8/window/mod.rs:420]: Fair enough :)

File: book/listings/todo/8/window/mod.rs (hunk 7 above)
Comment (zekefast) [reply book/listings/todo/8/window/mod.rs]: Fixed by 48572ed228fff0d426e56f1cfe409918ecf03d1b

**Comment by zekefast** [comment]
@Hofer-Julian Suggestions were applied.

**Review by Hofer-Julian — APPROVED** [review]

**Comment by zekefast** [comment]
Thanks!

**Comment by Hofer-Julian** [comment]
Thank you :D
